### ISSUE TITLE
Deserialization of Keypair with Serde is broken

### ISSUE THREAD

**Issue by jamesmunns** [dalek-cryptography/ed25519-dalek#111]
Hey there,

I am trying to use this to serialize and deserialize a `Keypair` item. Serialization seems to work, however it fails to deserialize. Here's a short example that shows this with `serde_json`, `toml`, and `postcard` (a binary format):

```rust
use ed25519_dalek::Keypair;
use serde::{Deserialize, Serialize};
use rand::rngs::OsRng;
use toml;
use serde_json;
use postcard;

#[derive(Debug, Serialize, Deserialize)]
struct Demo {
    keypair: Keypair
}

fn main() {
    println!("Make a key...");

    let demo = Demo { keypair: Keypair::generate(&mut OsRng) };

    println!("\n\nWrite to toml");
    let demo_toml = toml::to_string(&demo).unwrap();
    println!("{}", demo_toml);
    let demo_toml_rebuild: Result<Demo, _> = toml::from_str(&demo_toml);
    println!("{:?}", demo_toml_rebuild);

    println!("\n\nWrite to json");
    let demo_json = serde_json::to_string(&demo).unwrap();
    println!("{}", demo_json);
    let demo_json_rebuild: Result<Demo, _> = serde_json::from_str(&demo_json);
    println!("{:?}", demo_json_rebuild);

    println!("\n\nWrite to postcard");
    let demo_postcard = postcard::to_stdvec(&demo).unwrap();
    println!("{:?}", demo_postcard);
    let demo_postcard_rebuild: Result<Demo, _> = postcard::from_bytes(&demo_postcard);
    println!("{:?}", demo_postcard_rebuild);
}
```

Result:

```
Make a key...

Write to toml
keypair = [103, 204, 156, 181, 222, 111, 170, 252, 203, 212, 196, 168, 70, 26, 213, 243, 233, 53, 57, 8, 206, 225, 35, 124, 56, 77, 142, 204, 29, 57, 188, 149, 192, 244, 71, 62, 116, 74, 38, 91, 132, 215, 236, 229, 208, 89, 150, 52, 14, 182, 180, 242, 67, 51, 151, 41, 239, 68, 30, 40, 56, 38, 145, 192]

Err(Error { inner: ErrorInner { kind: Custom, line: Some(0), col: 0, at: Some(0), message: "invalid type: sequence, expected An ed25519 keypair, 64 bytes in total where the secret key is the first 32 bytes and is in unexpanded form, and the second 32 bytes is a compressed point for a public key.", key: ["keypair"] } })


Write to json
{"keypair":[103,204,156,181,222,111,170,252,203,212,196,168,70,26,213,243,233,53,57,8,206,225,35,124,56,77,142,204,29,57,188,149,192,244,71,62,116,74,38,91,132,215,236,229,208,89,150,52,14,182,180,242,67,51,151,41,239,68,30,40,56,38,145,192]}
Err(Error("invalid type: sequence, expected An ed25519 keypair, 64 bytes in total where the secret key is the first 32 bytes and is in unexpanded form, and the second 32 bytes is a compressed point for a public key.", line: 1, column: 12))


Write to postcard
[103, 204, 156, 181, 222, 111, 170, 252, 203, 212, 196, 168, 70, 26, 213, 243, 233, 53, 57, 8, 206, 225, 35, 124, 56, 77, 142, 204, 29, 57, 188, 149, 192, 244, 71, 62, 116, 74, 38, 91, 132, 215, 236, 229, 208, 89, 150, 52, 14, 182, 180, 242, 67, 51, 151, 41, 239, 68, 30, 40, 56, 38, 145, 192]
Err(DeserializeUnexpectedEnd)
```

I'll dig into this a bit more this week hopefully.

**Comment by jamesmunns** [comment]
This is specifically with `ed25519-dalek = { version = "1.0.0-pre.3", features = ["serde"] }`.

CC @tarcieri as I saw you mentioned broken serde tests in #102

**Comment by tarcieri** [comment]
That's a red herring, as that PR is to move to types from the [`ed25519` crate](https://crates.io/crates/ed25519), which provide their own serde serializers/deserializers.

**Comment by rainhead** [comment]
Signature and PublicKey are also affected.

**Comment by isislovecruft** [comment]
Merged #112 and created a new issue for @rainhead's comment.
